src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -1129,7 +1129,16 @@ public abstract static class AnyInstanceVariableNode extends PrimitiveArrayArgum
 
         @Specialization(limit = "getDynamicObjectCacheLimit()")
         protected boolean any(RubyDynamicObject self,
-                @CachedLibrary("self") DynamicObjectLibrary objectLibrary) {
+                @CachedLibrary("self") DynamicObjectLibrary objectLibrary,
+                @Cached BranchProfile hasPropertiesProfile) {
+            var shape = objectLibrary.getShape(self);
+
+            if (shape.getPropertyCount() == 0) {
+                return false;
+            }
+
+            hasPropertiesProfile.enter();
```

Comment:
Could you use a (binary) ConditionProfile here? So it profiles both branches, no matter which happens.